<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Planaria single-cell &mdash; pyemb  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../_static/documentation_options.js?v=5929fcd5"></script>
        <script src="../_static/doctools.js?v=9a2dae69"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="API" href="../api.html" />
    <link rel="prev" title="20Newsgroup documents" href="newsgroup.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            pyemb
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../inst_over.html">Installation and Overview</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Tutorials</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="mimic.html">MIMIC-III</a></li>
<li class="toctree-l2"><a class="reference internal" href="lyon.html">School contact graph</a></li>
<li class="toctree-l2"><a class="reference internal" href="newsgroup.html">20Newsgroup documents</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Planaria single-cell</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#load-data">Load data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#dimension-selection-and-visualisation">Dimension selection and visualisation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#construct-tree">Construct tree</a></li>
<li class="toctree-l3"><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="../refs.html">Contributors and References</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">pyemb</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="index.html">Tutorials</a></li>
      <li class="breadcrumb-item active">Planaria single-cell</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/tutorials/planaria.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="planaria-single-cell">
<h1>Planaria single-cell<a class="headerlink" href="#planaria-single-cell" title="Link to this heading"></a></h1>
<p>This section uses <a class="reference external" href="https://shiny.mdc-berlin.de/psca/">data</a> from the
paper <a class="reference external" href="https://www.science.org/doi/abs/10.1126/science.aaq1723">Plass et al.,
2018</a>. It
contains expression levels of <span class="math notranslate nohighlight">\(p= 5821\)</span> genes measured across
<span class="math notranslate nohighlight">\(n=5000\)</span> individual cells from adult planarians, a type of
flatworm. Such data offer the possibility of discovering the cell
lineage tree of an entire animal: the aim is to find out if the data
reflect the tree-structured process by which stem cells differentiate
into a variety of distinct cell types. These data were prepared using
the Python package
<a class="reference external" href="https://scanpy.readthedocs.io/en/stable/index.html">Scanpy</a>, where
the script on how to do this can be found
<a class="reference external" href="https://nbviewer.org/github/rajewsky-lab/planarian_lineages/blob/master/paga/preprocessing.ipynb">here</a>.</p>
<section id="load-data">
<h2>Load data<a class="headerlink" href="#load-data" title="Link to this heading"></a></h2>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span> <span class="o">=</span> <span class="n">eb</span><span class="o">.</span><span class="n">load_planaria</span><span class="p">()</span>

<span class="n">Y</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;Y&#39;</span><span class="p">]</span>
<span class="n">labels</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;labels&#39;</span><span class="p">]</span>
<span class="n">ordered_cats</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;cell_types&#39;</span><span class="p">]</span>
<span class="n">colors</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;color_dict&#39;</span><span class="p">]</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Data</span> <span class="n">loaded</span> <span class="n">successfully</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="p">(</span><span class="n">n</span><span class="p">,</span><span class="n">p</span><span class="p">)</span> <span class="o">=</span> <span class="n">Y</span><span class="o">.</span><span class="n">shape</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Data matrix is&#39;</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="s1">&#39;samples by&#39;</span><span class="p">,</span> <span class="n">p</span><span class="p">,</span> <span class="s1">&#39;features&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Data</span> <span class="n">matrix</span> <span class="ow">is</span> <span class="mi">5000</span> <span class="n">samples</span> <span class="n">by</span> <span class="mi">100</span> <span class="n">features</span>
</pre></div>
</div>
</section>
<section id="dimension-selection-and-visualisation">
<h2>Dimension selection and visualisation<a class="headerlink" href="#dimension-selection-and-visualisation" title="Link to this heading"></a></h2>
<p>After selecting the dimension <a class="reference external" href="https://arxiv.org/pdf/2208.11665">Whiteley et al.,
2022</a>, perform PCA with embedding
function and visualise.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ws</span><span class="p">,</span> <span class="n">dim</span> <span class="o">=</span> <span class="n">eb</span><span class="o">.</span><span class="n">wasserstein_dimension_select</span><span class="p">(</span><span class="n">Y</span><span class="p">,</span> <span class="nb">range</span><span class="p">(</span><span class="mi">40</span><span class="p">),</span> <span class="n">split</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Selected dimension: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">dim</span><span class="p">))</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>100%|██████████| 40/40 [01:06&lt;00:00,  1.67s/it]
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Recommended</span> <span class="n">dimension</span><span class="p">:</span> <span class="mi">16</span><span class="p">,</span> <span class="n">Wasserstein</span> <span class="n">distance</span> <span class="mf">0.18280</span>
<span class="n">Selected</span> <span class="n">dimension</span><span class="p">:</span> <span class="mi">16</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">zeta</span> <span class="o">=</span> <span class="n">eb</span><span class="o">.</span><span class="n">embed</span><span class="p">(</span><span class="n">Y</span><span class="p">,</span> <span class="n">d</span><span class="o">=</span><span class="mi">14</span><span class="p">,</span> <span class="n">version</span><span class="o">=</span><span class="s1">&#39;full&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## TSNE</span>
<span class="kn">from</span> <span class="nn">sklearn.manifold</span> <span class="kn">import</span> <span class="n">TSNE</span>
<span class="n">tsne</span> <span class="o">=</span> <span class="n">TSNE</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">perplexity</span> <span class="o">=</span> <span class="mi">30</span><span class="p">)</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">zeta</span><span class="p">)</span>
</pre></div>
</div>
<p>We can plot the two representations of our data with the
<code class="docutils literal notranslate"><span class="pre">snapshot_plot</span></code> function.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">pca_fig</span> <span class="o">=</span> <span class="n">eb</span><span class="o">.</span><span class="n">snapshot_plot</span><span class="p">(</span>
    <span class="n">embedding</span> <span class="o">=</span> <span class="p">[</span><span class="n">zeta</span><span class="p">[:,:</span><span class="mi">2</span><span class="p">],</span><span class="n">tsne</span><span class="p">],</span>
    <span class="n">node_labels</span> <span class="o">=</span> <span class="n">labels</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span>
    <span class="n">c</span> <span class="o">=</span> <span class="n">colors</span><span class="p">,</span>
    <span class="n">title</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;PCA&#39;</span><span class="p">,</span><span class="s1">&#39;tSNE&#39;</span><span class="p">],</span>
    <span class="n">add_legend</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">max_legend_cols</span> <span class="o">=</span> <span class="mi">6</span><span class="p">,</span>
   <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="mi">15</span><span class="p">,</span><span class="mf">7.5</span><span class="p">),</span>
   <span class="n">move_legend</span><span class="o">=</span> <span class="p">(</span><span class="mf">.5</span><span class="p">,</span><span class="o">-</span><span class="mf">.3</span><span class="p">),</span>
    <span class="c1"># Apply other matplotlib settings</span>
    <span class="n">s</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
</pre></div>
</div>
<img alt="../_images/planaria_11_0.png" src="../_images/planaria_11_0.png" />
</section>
<section id="construct-tree">
<h2>Construct tree<a class="headerlink" href="#construct-tree" title="Link to this heading"></a></h2>
<p>From here we want to perform hierarchical clustering on the data and
simplify this tree. This can be done using the hierarchical clustering
module.</p>
<p>We use hierarchical clustering with dot products as descibed in <a class="reference external" href="https://proceedings.neurips.cc/paper_files/paper/2023/file/6521937507d78f327cd402401be73bf2-Paper-Conference.pdf">Gray et
al.,
2024</a>.
This is the default HC in the class <code class="docutils literal notranslate"><span class="pre">ConstructTree</span></code> so we can just
give it our point cloud. Otherwise, the HC can be done first and the
model can be given to the <code class="docutils literal notranslate"><span class="pre">ConstructTree</span></code> class.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">epsilon</span></code> parameter controls the threshold for condensing tree and
if set to zero the full tree will be given.</p>
<p>This can then be plotted. Points are coloured by labels, if a plotted
node is a collection of data points then <code class="docutils literal notranslate"><span class="pre">colour_threshold</span></code> controls
when to colour this node by the majority type of data point (colour by
majority if proportion of majority is greater than colour_threshold),
else, if there is no majority node, it is plotted white.</p>
<p>Layouts, node settings and others can also be changed. See the
documentation for details of this.</p>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tree</span> <span class="o">=</span> <span class="n">eb</span><span class="o">.</span><span class="n">ConstructTree</span><span class="p">(</span><span class="n">zeta</span><span class="p">,</span> <span class="n">epsilon</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>
<span class="n">tree</span><span class="o">.</span><span class="n">fit</span><span class="p">()</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Performing</span> <span class="n">clustering</span><span class="o">...</span>
<span class="n">Calculating</span> <span class="n">branch</span> <span class="n">lengths</span><span class="o">...</span>
<span class="n">Constructing</span> <span class="n">tree</span><span class="o">...</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&lt;</span><span class="n">pyemb</span><span class="o">.</span><span class="n">hc</span><span class="o">.</span><span class="n">ConstructTree</span> <span class="n">at</span> <span class="mh">0x7a1014496170</span><span class="o">&gt;</span>
</pre></div>
</div>
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">labels</span> <span class="o">=</span> <span class="n">labels</span><span class="p">,</span>
                <span class="n">colours</span> <span class="o">=</span> <span class="n">colors</span><span class="p">,</span>
                <span class="n">add_legend</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
                <span class="n">move_legend</span> <span class="o">=</span> <span class="p">(</span><span class="mf">.5</span><span class="p">,</span><span class="o">-</span><span class="mf">.25</span><span class="p">),</span>
                <span class="n">prog</span> <span class="o">=</span> <span class="s1">&#39;twopi&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>100%|██████████| 250/250 [00:03&lt;00:00, 69.38it/s]
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">BarnesHut</span> <span class="n">Approximation</span>  <span class="n">took</span>  <span class="mf">2.26</span>  <span class="n">seconds</span>
<span class="n">Repulsion</span> <span class="n">forces</span>  <span class="n">took</span>  <span class="mf">1.08</span>  <span class="n">seconds</span>
<span class="n">Gravitational</span> <span class="n">forces</span>  <span class="n">took</span>  <span class="mf">0.02</span>  <span class="n">seconds</span>
<span class="n">Attraction</span> <span class="n">forces</span>  <span class="n">took</span>  <span class="mf">0.01</span>  <span class="n">seconds</span>
<span class="n">AdjustSpeedAndApplyForces</span> <span class="n">step</span>  <span class="n">took</span>  <span class="mf">0.12</span>  <span class="n">seconds</span>
</pre></div>
</div>
<img alt="../_images/planaria_18_2.png" src="../_images/planaria_18_2.png" />
</section>
<section id="references">
<h2>References<a class="headerlink" href="#references" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Plass, M., Solana, J., Wolf, F.A., Ayoub, S., Misios, A., Glažar, P.,
Obermayer, B., Theis, F.J., Kocks, C. and Rajewsky, N., 2018. Cell
type atlas and lineage tree of a whole complex animal by single-cell
transcriptomics. Science, 360(6391), p.eaaq1723.</p></li>
<li><p>Whiteley, N., Gray, A. and Rubin-Delanchy, P., 2022. Statistical
exploration of the Manifold Hypothesis.</p></li>
<li><p>Gray, A., Modell, A., Rubin-Delanchy, P. and Whiteley, N., 2024.
Hierarchical clustering with dot products recovers hidden tree
structure. Advances in Neural Information Processing Systems, 36.</p></li>
</ul>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="newsgroup.html" class="btn btn-neutral float-left" title="20Newsgroup documents" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../api.html" class="btn btn-neutral float-right" title="API" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Annie Gray.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>